File: start_server.js
const http = require('http');
const fs = require('fs');
const path = require('path');
const { exec } = require('child_process');
const { escape } = require('html-escaper');

class CustomHTTPRequestHandler {
  constructor(req, res) {
    this.req = req;
    this.res = res;
    this.directory = process.cwd();
  }

  handleRequest() {
    if (this.req.url === '/') {
      this.checkAndGenerateReport();
      this.serveFile('repository_files.html');
    } else if (this.req.url.endsWith('.txt')) {
      this.serveTextFile();
    } else {
      this.serveFile(this.req.url.slice(1));
    }
  }

  checkAndGenerateReport() {
    const htmlFile = path.join(this.directory, 'repository_files.html');
    fs.stat(htmlFile, (err, stats) => {
      if (err || Date.now() - stats.mtime.getTime() > 60000) {
        console.log(`${htmlFile} is older than 60 seconds or does not exist. Running generate_report.js...`);
        exec('node generate_report.js', (error, stdout, stderr) => {
          if (error) {
            console.error(`Error: ${error.message}`);
            return;
          }
          if (stderr) {
            console.error(`Error: ${stderr}`);
            return;
          }
          console.log(`Output: ${stdout}`);
        });
      } else {
        console.log(`${htmlFile} is up-to-date.`);
      }
    });
  }

  serveTextFile() {
    const filePath = path.join(this.directory, this.req.url.slice(1));
    fs.readFile(filePath, 'utf8', (err, content) => {
      if (err) {
        this.sendError(404, 'File not found');
        return;
      }

      const htmlContent = `
        <!DOCTYPE html>
        <html lang="en">
        <head>
            <meta charset="UTF-8">
            <meta name="viewport" content="width=device-width, initial-scale=1.0">
            <title>${path.basename(filePath)}</title>
            <style>
                body { font-family: Arial, sans-serif; line-height: 1.6; padding: 20px; }
                pre { background-color: #f4f4f4; padding: 15px; border-radius: 5px; white-space: pre-wrap; word-wrap: break-word; }
            </style>
        </head>
        <body>
            <h1>${path.basename(filePath)}</h1>
            <pre>${escape(content)}</pre>
        </body>
        </html>
      `;

      this.res.writeHead(200, { 'Content-Type': 'text/html; charset=utf-8' });
      this.res.end(htmlContent);
    });
  }

  serveFile(fileName) {
    const filePath = path.join(this.directory, fileName);
    fs.readFile(filePath, (err, content) => {
      if (err) {
        this.sendError(404, 'File not found');
        return;
      }

      const ext = path.extname(filePath).toLowerCase();
      const contentType = {
        '.html': 'text/html',
        '.js': 'text/javascript',
        '.css': 'text/css',
        '.json': 'application/json',
        '.png': 'image/png',
        '.jpg': 'image/jpeg',
        '.gif': 'image/gif',
      }[ext] || 'application/octet-stream';

      this.res.writeHead(200, { 'Content-Type': contentType });
      this.res.end(content);
    });
  }

  sendError(statusCode, message) {
    this.res.writeHead(statusCode, { 'Content-Type': 'text/plain' });
    this.res.end(message);
  }
}

function runServer(port, directory) {
  process.chdir(directory);
  const server = http.createServer((req, res) => {
    const handler = new CustomHTTPRequestHandler(req, res);
    handler.handleRequest();
  });

  server.listen(port, () => {
    console.log(`Serving HTTP on 0.0.0.0 port ${port} (http://0.0.0.0:${port}/) ...`);
  });
}

function parseArguments() {
  const args = process.argv.slice(2);
  let port = 8000;
  let directory = process.cwd();

  for (let i = 0; i < args.length; i += 2) {
    if (args[i] === '-p' || args[i] === '--port') {
      port = parseInt(args[i + 1]);
    } else if (args[i] === '-d' || args[i] === '--directory') {
      directory = args[i + 1];
    }
  }

  return { port, directory };
}

const { port, directory } = parseArguments();
runServer(port, directory);

File: src/python/start_server.py
# start_server.py
# to run: python3 start_server.py

# start_server: v2

import http.server
import socketserver
import argparse
import os
import time
import subprocess
import html
import urllib.parse
from datetime import datetime
import random
import string


class CustomHTTPRequestHandler(http.server.SimpleHTTPRequestHandler):
    def __init__(self, *args, **kwargs):
        self.directory = kwargs.pop("directory", os.getcwd())
        super().__init__(*args, directory=self.directory, **kwargs)

    def do_GET(self):
        if self.path == "/":
            super().do_GET()
        elif self.path == "/log.html":
            self.check_and_generate_report()
            super().do_GET()
        elif self.path == "/chat.html":
            self.check_and_generate_chat_html()
            super().do_GET()
        elif self.path.endswith(".txt"):
            self.serve_text_file()
        else:
            super().do_GET()

    def do_POST(self):
        if self.path == "/chat.html":
            self.handle_chat_post()
        else:
            self.send_error(405, "Method Not Allowed")

    def handle_chat_post(self):
        content_length = int(self.headers["Content-Length"])
        post_data = self.rfile.read(content_length).decode("utf-8")
        params = urllib.parse.parse_qs(post_data)

        author = params.get("author", [""])[0]
        message = params.get("message", [""])[0]

        if author and message:
            self.save_message(author, message)
            self.send_response(200)
            self.send_header("Content-type", "text/html")
            self.end_headers()
            self.wfile.write(b"Message saved successfully")
            # commit the message to the git repository using commit_files.py
            subprocess.run(["python", "commit_files.py", "message"], check=True)
            # redirect back to chat.html
            self.wfile.write(b'<meta http-equiv="refresh" content="1;url=/chat.html">')

        else:
            self.send_error(400, "Bad Request: Missing author or message")

    def save_message(self, author, message):
        today = datetime.now().strftime("%Y-%m-%d")
        directory = os.path.join(self.directory, "message", today)
        os.makedirs(directory, exist_ok=True)

        title = self.generate_title(message)
        filename = f"{title}.txt"
        filepath = os.path.join(directory, filename)

        with open(filepath, "w", encoding="utf-8") as f:
            f.write(message)
            f.write(f"\n\nauthor: {author}")

    def generate_title(self, message):
        words = message.split()[:5]
        title = "_".join(words)
        title = "".join(c for c in title if c.isalnum() or c in ["_", "-"])
        if not title:
            title = "".join(random.choices(string.ascii_lowercase, k=10))
        return title

    def check_and_generate_report(self):
        html_file = os.path.join(self.directory, "log.html")
        if (
            not os.path.exists(html_file)
            or time.time() - os.path.getmtime(html_file) > 60
        ):
            print(
                f"{html_file} is older than 60 seconds or does not exist. Running generate_report.py..."
            )
            subprocess.run(["python", "src/python/generate_report.py"], check=True)
        else:
            print(f"{html_file} is up-to-date.")

    def check_and_generate_chat_html(self):
        chat_html_file = os.path.join(self.directory, "chat.html")
        if (
            not os.path.exists(chat_html_file)
            or time.time() - os.path.getmtime(chat_html_file) > 60
        ):
            print(
                f"{chat_html_file} is older than 60 seconds or does not exist. Running generate_chat_html.py..."
            )
            subprocess.run(["python", "src/python/generate_chat_html.py"], check=True)
        else:
            print(f"{chat_html_file} is up-to-date.")

    def serve_text_file(self):
        path = self.translate_path(self.path)
        try:
            with open(path, "r", encoding="utf-8") as f:
                content = f.read()

            self.send_response(200)
            self.send_header("Content-type", "text/html; charset=utf-8")
            self.end_headers()

            html_content = f"""
            <!DOCTYPE html>
            <html lang="en">
            <head>
                <meta charset="UTF-8">
                <meta name="viewport" content="width=device-width, initial-scale=1.0">
                <title>{os.path.basename(path)}</title>
                <style>
                    body {{ font-family: Arial, sans-serif; line-height: 1.6; padding: 20px; }}
                    pre {{ background-color: #f4f4f4; padding: 15px; border-radius: 5px; white-space: pre-wrap; word-wrap: break-word; }}
                </style>
            </head>
            <body>
                <h1>{os.path.basename(path)}</h1>
                <pre>{html.escape(content)}</pre>
            </body>
            </html>
            """

            self.wfile.write(html_content.encode("utf-8"))
        except IOError:
            self.send_error(404, "File not found")


def run_server(port, directory):
    os.chdir(directory)
    handler = CustomHTTPRequestHandler
    httpd = socketserver.TCPServer(("", port), handler)
    try:
         print(f"Serving HTTP on 0.0.0.0 port {port} (http://0.0.0.0:{port}/) ...")
         httpd.serve_forever()
    except KeyboardInterrupt:
        pass
    httpd.server_close()


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Run a simple HTTP server.")
    parser.add_argument(
        "-p", "--port", type=int, default=8000, help="Port to serve on (default: 8000)"
    )
    parser.add_argument(
        "-d",
        "--directory",
        type=str,
        default=os.getcwd(),
        help="Directory to serve (default: current directory)",
    )

    args = parser.parse_args()

    run_server(args.port, args.directory)
